src/Microcosm.js

```diff
@@ -114,26 +111,13 @@ Microcosm.prototype = {
       throw TypeError(`Tried to push ${ action }, but is not a function.`)
     }
 
-    let transaction = { action, done: false }
-
-    this.transactions.push(transaction)
-
-    return signal((error, body, done) => {
-      if (error) {
-        transaction.done  = true
-        transaction.error = error
-      }
+    let transaction = new Transaction(action, params)
 
-      if (!transaction.done) {
-        transaction.body = body
-      }
+    transaction.listen(this.rollforward.bind(this))
 
-      transaction.done = transaction.done || done
-
-      this.rollforward()
+    this.transactions.push(transaction)
 
-      return body
-    }, action.apply(this, params))
+    return transaction.run(params)
   },
```

Comment:
Actually this still feels a bit awkward. Transactions are event emitters, but they only have one event. I had considered just doing:

`new Transaction(action, params, this.rollforward.bind(this)`